File: Python/Easy/MaxProfit.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        profit = 0

        left, right = 0, 1

        while right < len(prices):

            # right pointer at lowest price
            if prices[left] > prices[right]:
                left = right
            else:
                profit = max(profit, prices[right] - prices[left])
            
            right += 1
            
        return profit

s = Solution()

s.maxProfit([7,1,5,3,6,4])

File: CPP/Easy/MaxProfit.cpp
#include <vector>

using namespace std;

/**
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int profit = 0;

        int left = 0, right = 1;

        while (right < prices.size())
        {
            // right pointer at lowest price
            if (prices[left] > prices[right])
                left = right;
            else
                profit = max(profit, prices[right] - prices[left]);

            right++;
        }

        return profit;
    }
};